go/util/jwt/jwt.go

```diff
@@ -82,48 +109,111 @@ func (p Permissions) Validate() error {
 
 		providers[permission.Provider.String()] = true
 
-		if permission.OSeq > 0 && permission.GSeq == 0 {
-			return fmt.Errorf("%w: gseq must be set if oseq is set", ErrJWTValidation)
-		}
-
-		if permission.GSeq > 0 && permission.DSeq == 0 {
-			return fmt.Errorf("%w: dseq must be set if gseq is set", ErrJWTValidation)
-		}
-
-		if len(permission.Scope) == 0 {
-			return fmt.Errorf("%w: empty permissions scope", ErrJWTValidation)
-		}
+		switch permission.Access {
+		case AccessTypeFull:
+			if len(permission.Deployments) > 0 {
+				return fmt.Errorf("%w: permission with full access cannot have deployments set", ErrJWTValidation)
+			}
+			if len(permission.Scope) > 0 {
+				return fmt.Errorf("%w: permission with full access cannot have scope set", ErrJWTValidation)
+			}
+		case AccessTypeScoped:
+			if permission.Deployments != nil || len(permission.Deployments) > 0 {
+				return fmt.Errorf("%w: permission with scoped access cannot have deployments set", ErrJWTValidation)
+			}
+			if len(permission.Scope) == 0 {
+				return fmt.Errorf("%w: permission with scope access must have scope set", ErrJWTValidation)
+			}
 
-		scopes := make(map[string]bool)
+			if err := permission.Scope.Validate(); err != nil {
+				return err
+			}
 
-		for _, scope := range permission.Scope {
-			if _, ok := validScopes[scope.String()]; !ok {
-				return fmt.Errorf("%w: invalid scope %s", ErrJWTValidation, scope)
+		case AccessTypeGranular:
+			if permission.Scope != nil || len(permission.Scope) > 0 {
+				return fmt.Errorf("%w: permission with granular access cannot have scope set", ErrJWTValidation)
 			}
 
-			if _, ok := scopes[scope.String()]; ok {
-				return fmt.Errorf("%w: duplicate scope %s", ErrJWTValidation, scope)
+			if len(permission.Deployments) == 0 {
+				return fmt.Errorf("%w: granular access must have permissions set", ErrJWTValidation)
 			}
 
-			scopes[scope.String()] = true
-		}
+			for _, deployment := range permission.Deployments {
+				if len(deployment.Scope) == 0 {
+					return fmt.Errorf("%w: deployment permission scope must be set", ErrJWTValidation)
+				}
+
+				if deployment.DSeq == 0 {
+					return fmt.Errorf("%w: dseq must be set", ErrJWTValidation)
+				}
+
+				if deployment.OSeq > 0 && deployment.GSeq == 0 {
+					return fmt.Errorf("%w: gseq must be set if oseq is set", ErrJWTValidation)
+				}
+
+				if len(deployment.Scope) == 0 {
+					return fmt.Errorf("%w: empty permissions scope", ErrJWTValidation)
+				}
+
+				if err := deployment.Scope.Validate(); err != nil {
+					return err
+				}
+
+				if len(deployment.Services) == 0 {
+					return fmt.Errorf("%w: empty services", ErrJWTValidation)
+				}
+
```

Comment:
Suggestion: make these validations into their own functions with a descriptive name for easy reading and updates whenever necessary.

Reply:
i don't think it is necessary here:
1. those validations are one logical piece according to a spec
2. making them into dedicated functions will brake first glance logic and force developer to check each and every function